feat(net): add vim.net.request() and :e integration

tampueroc opened the pull request:
This PR introduces `vim.net.request()` as a new minimal Lua API for fetching content from network URLs and integrates directly into Neovim `:e` command for remote file opening. Closes #23232 .

# Implementation details
- Defines `vim.net.request(url, opts, on_response?)` in `runtime/lua/vim/net.lua.`
- Uses `vim.system()` to invoke curl with:
    - `--silent`, `--show-error`, `--fail`
    - `--location` (follow redirects)
    - `--retry` <n> (default 3)
- Supports an `opts.outpath` path to stream directly to a file
- Runs asynchronously and always calls `on_response(err, response)`.  
- When`opts.outpath` is set, the value passed as the `response` body to `on_response` callback is a boolean `true`

# Proposed API Usage
1. Asynchronous request
```lua
vim.net.request("https://httpbingo.org/get", { retry = 2 }, function(err, response)
  if err then
    print("Fetch failed:", err)
  else
    print("Got body of length:", #response.body)
  end
end)
```
2. Download to file
```lua
vim.net.request("https://httpbingo.org/get", { outpath = "out_async.txt" }, function(err)
  if err then print("Error:", err) end
end)
```
3. Remote :e integration (in runtime/plugin/net.lua) fetches into buffer
```vim
:e https://httpbingo.org/get
```

# Next steps
- Add support for custom headers
- Consider a tiny `vim.net.download()` convenience wrapper if users need it
- Return a request handle which can be operated on (see [comment](https://github.com/neovim/neovim/pull/34140#discussion_r2173754322))
- Expose more `curl` flags through `opts`

clason commented:
Let's please keep simple download and full requests separate. I'd prefer to finish the former before tackling the latter unless it's absolutely necessary to implement them together (which I don't see).

tampueroc commented:
> Let's please keep simple download and full requests separate. I'd prefer to finish the former before tackling the latter unless it's absolutely necessary to implement them together (which I don't see).

That's my intention exactly! As we discussed in the minimal PR for `vim.net.download()` #33964 I'm fully aligned that we need to keep these efforts seprate.  I opened this PR as a draft specifically so we can align the vim.net.request() scope and implementation and this way I can address early feedback. As soon as the `vim.net.download()` PR is merged I will rebase this branch as to remove the noise and ensure its a clean focused PR.

I completely agree with and appreciate the Bell System-style strategy for this initiative. I understand that previous efforts struggled due to complexity, so I'm really trying to avoid that for future development.

tampueroc commented:
FYI Core functionality is implemented but functional tests for new API function is in progress. 
In the meantime can I have some feedback on why this tests might be failing? Locally it doesn't fail. @lewis6991 

```
runtime/lua/vim/_extui/messages.lua:229:13 [Warning] Can not infer type. (no-unknown)
          local end_col = #mid + col
                ^^^^^^^
runtime/lua/vim/_extui/messages.lua:235:15 [Warning] Can not infer type. (no-unknown)
            local ecol = math.min(end_col, start_line and #start_line or -1)
                  ^^^^
runtime/lua/vim/_extui/messages.lua:254:13 [Warning] Can not infer type. (no-unknown)
          head, col = '', end_col
                ^^^
```

lewis6991 commented on runtime/lua/vim/net.lua:
```diff
@@ -0,0 +1,128 @@
+--- @param on_exit? fun(err?: string, content?: string) Optional callback. If omitted, runs synchronously.
+---
+--- @return string|nil, string? On success (sync): content; On failure (sync) nil and error message.
+function M.request(url, opts, on_exit)
```
Even though the API is separate functions, I would pull out the common code between the two into a local function.

justinmk replied:
why are they separate? the interfaces are nearly the same

justinmk replied:
I don't see why the return types should be different.

Not trying to be stubborn and I agree it's not a hill to die on.

My suggestion is to start with a basic implementation of `vim.net.fetch()` and see if we actually need a `vim.net.download()`. If "collect the streamed content and save it to a file" is too cumbersome, then yeah maybe a different function is needed. But that can be a later step.

justinmk commented:
> Let's please keep simple download and full requests separate.

Why are they separate? What is `vim.net.download()` doing that requires it to be separate from `fetch()` ?

justinmk commented on runtime/pack/dist/opt/netrw/plugin/netrwPlugin.vim:
```diff
@@ -38,7 +38,7 @@ augroup END
     au!
     au BufReadCmd file://* call netrw#FileUrlEdit(expand("<amatch>"))
-    au BufReadCmd ftp://*,rcp://*,scp://*,http://*,https://*,dav://*,davs://*,rsync://*,sftp://* exe "sil doau BufReadPre ".fnameescape(expand("<amatch>"))|call netrw#Nread(2,expand("<amatch>"))|exe "sil doau BufReadPost ".fnameescape(expand("<amatch>"))
+    au BufReadCmd ftp://*,rcp://*,scp://*,dav://*,davs://*,rsync://*,sftp://* exe "sil doau BufReadPre ".fnameescape(expand("<amatch>"))|call netrw#Nread(2,expand("<amatch>"))|exe "sil doau BufReadPost ".fnameescape(expand("<amatch>"))
```
👍 this seems like a good way to incrementally do stuff

clason commented:
>  fetch() aka request() aka download()

Because these are **not** "aka". Requests are _much_ more general (`POST`!) than a simple `fetch`/`download`, and not having to wrangle request parameters for simply downloading a spell file is worth a (simple) convenience function.

tampueroc commented:
What is the final resolution regarding the separate functions for the `net` module? @clason @justinmk 
I'm happy to close the other pull request for `vim.net.download()` (#33964), incorporate the explicit download behaviour under `vim.net.request()` and extend the functions behaviour in future PRs if and only if is neccesary.

justinmk commented:
@tampueroc to unblock you, my suggestion is

- Continue with 1 PR for the low-level `request()`, and let's see what it actually looks like.
- Do not attempt to support tons of curl features, just support the bare minimum that is needed to download a file. 
    - But keep the door open to add *more* curl features in the future.

If using this new function to download a file feels cumbersome, *then* we can either add a parameter to it that activates "download to file" mode, or add a simple `download()` function that *wraps* `request()`.

clason commented on runtime/lua/vim/health/health.lua:
```diff
@@ -406,6 +406,56 @@ local function check_external_tools()
+      health.warn('curl is installed but failed to run `curl --version`', { curl_out })
+    end
+  else
+    health.error('curl not found', {
```
Do we want to make this an error or a warning? What's the official stance -- is `curl` now a _required_ runtime dependency?

justinmk replied:
it has been "required" since the beginning for e.g. spell and netrw. 

if we actually get reports from people that dont have curl and are bothered by the error, we can change to a warning.

meanwhile, I think error is probably the right designation because it's definitely something that needs attention from the user.

clason commented on runtime/plugin/net.lua:
```diff
@@ -0,0 +1,43 @@
+end
+
+-- Autocmd group for remote URLs
+local aug = vim.api.nvim_create_augroup('NetRemoteFile', { clear = true })
```
```suggestion
local aug = vim.api.nvim_create_augroup('NetRemoteFile', { clear = true })
```
`clear = true` is the default, so you can just put `group = vim.api.nvim_create_augroup('NetRemoteFile', {})` in the `create_autocmd` call. (You only need to bind it if you want to reuse it multiple times.)

justinmk commented on runtime/lua/vim/health/health.lua:
```diff
@@ -406,6 +406,56 @@ local function check_external_tools()
   end
+  if vim.fn.executable('curl') == 1 then
+    local curl_path = vim.fn.exepath('curl')
+    local curl_out = vim.fn.system({ curl_path, '--version' })
```
use `vim.system()`. that avoids fragile things like `vim.v.shell_error` :)

tampueroc replied:
Thank you for the suggestion!

justinmk commented on runtime/lua/vim/health/health.lua:
```diff
@@ -406,6 +406,56 @@ local function check_external_tools()
+        local val = vim.env[var] or vim.env[var:upper()]
+        if val then
+          if not added_env_header then
+            table.insert(lines, 'curl-related environment variables:')
```
great idea, this kind of visibility is important

justinmk commented on runtime/lua/vim/net.lua:
```diff
@@ -0,0 +1,78 @@
+
+  -- Async mode
+  if on_exit then
+    vim.system(args, { text = true }, function(res)
```
nit: for sync mode would it make sense for the logic to define a `on_exit` and then pass it to `vim.system` ? that lets us have 1 codepath

tampueroc replied:
Thank you!

justinmk commented on runtime/plugin/net.lua:
move to `runtime/plugin/nvim/net.lua`

all builtin plugins should live in a "nvim" namespace. we shouldn't occupy a top-level name every time we add a builtin plugin.

clason replied:
> This matters for e.g. vimscript net#Foo() vs nvim#net#Foo(). 

No, it doesn't, because that is only relevant for `/autoload` (and `/lua`), _not_ `/plugin`.

justinmk replied:
It's relevant for plugin/ autoload/, etc, because there is no other way to reason about which files "belong" to a plugin namespace. https://github.com/neovim/neovim/issues/34704

If `plugin/net.x` exists, the plugin "name" is "net". If `plugin/nvim/net.x` exists, the plugin "name" is "nvim".

justinmk commented on runtime/plugin/net.lua:
```diff
@@ -0,0 +1,43 @@
+vim.api.nvim_create_autocmd('BufReadCmd', {
+  group = vim.api.nvim_create_augroup('NetRemoteFile', {}),
+  pattern = { 'http://*', 'https://*' },
+  desc = 'Load remote files via vim.net.request()',
```
no need to mention implementation details here
```suggestion
  desc = 'Edit remote files (:edit https://example.com)',
```

justinmk reviewed (approved):
This looks like a good start to me. I will review again asap to check if there are any one-way doors that we need to decide now, at the interface level

justinmk commented on runtime/doc/lua.txt:
```diff
@@ -4656,4 +4656,30 @@ tohtml.tohtml({winid}, {opt})                         *tohtml.tohtml.tohtml()*
+                     failures (default: 3).
+                   • output (string|nil): If set, path to save response body
+                     instead of returning it.
+      • {on_exit}  (`fun(err?: string, content?: string)?`)
```
does `on_response` make sense as the name for this? "exit" is an implementation detail.  the fact that curl is used internally isn't relevant and shouldn't be surfaced in the request() interface (for now).

tampueroc replied:
Agree!

tampueroc commented on runtime/lua/vim/health/health.lua:
```diff
@@ -406,6 +406,56 @@ local function check_external_tools()
+
+    if vim.v.shell_error == 0 then
+      local lines = {}
+      table.insert(lines, string.format('%s (%s)', curl_version, curl_path))
```
Here are the curl versions for every command we use in the implementation ([1](https://curl.se/docs/optionsall.html)):
 | command | version | year of release
 | ----------- | ----------- |----------- |
|--silent (-s) | 4.0 | ~1998
--fail (-f) | 4.0 | ~1998
--output (-o) | 4.0 | ~1998
--location (-L) | 4.9 | ~1998
--show-error (-S) | 5.9 | ~1999
--retry | 7.12 .3 | ~2004

We should be safe triggering a warning if lower than `7.12.3`

(1): [https://curl.se/docs/optionsall.html](https://curl.se/docs/optionsall.html)

lewis6991 commented on runtime/doc/lua.txt:
```diff
@@ -4656,4 +4656,36 @@ tohtml.tohtml({winid}, {opt})                         *tohtml.tohtml.tohtml()*
+    This function operates in two modes:
+    • Synchronous (blocking): If `on_response` is omitted, it waits for the
+      request to complete and returns the result.
+    • Asynchronous (non-blocking): If `on_response` is provided, it returns
```
One door that needs to remain open is being able to cancel requests in the future. That means the async call will need to return some handle which can be operated on.

tampueroc replied:
How about returning the underlying `vim.system()` job from the async call, that way callers can use the full `vim.SystemObj` API e.g `:kill()` to cancel requests?

lewis6991 replied:
Perfect.

justinmk commented on runtime/lua/vim/net.lua:
```diff
@@ -0,0 +1,71 @@
+---   - `retry`   (integer|nil): Number of retries on transient failures (default: 3).
+---   - `output`  (string|nil): A file path to save the response body to. If set, the success value will be `true` instead of the response body.
+--- @param on_response? fun(err?: string, content?: string|boolean) Optional callback for async execution.
+---   It is invoked with `(err, nil)` on failure or `(nil, content)` on success.
```
This appears to match the convention of `:help lsp-handler` and luv (`vim.uv`) callbacks 👍

justinmk commented on runtime/lua/vim/net.lua:
```diff
@@ -0,0 +1,71 @@
+---   - `output`  (string|nil): A file path to save the response body to. If set, the success value will be `true` instead of the response body.
+--- @param on_response? fun(err?: string, content?: string|boolean) Optional callback for async execution.
+---   It is invoked with `(err, nil)` on failure or `(nil, content)` on success.
+--- @return string|boolean|nil In sync mode, returns the response body or `true` on success; otherwise `nil`
```
the wording is kind of confusing, does this make sense:
```suggestion
--- @return string|boolean|nil Response body (sync mode) or `true` (async) on success; `nil` on failure
```

lewis6991 commented on runtime/lua/vim/net.lua:
```diff
@@ -0,0 +1,71 @@
+--- @param opts? table Optional parameters:
+---   - `verbose` (boolean|nil): Enables curl verbose output.
+---   - `retry`   (integer|nil): Number of retries on transient failures (default: 3).
+---   - `output`  (string|nil): A file path to save the response body to. If set, the success value will be `true` instead of the response body.
```
I think `output` or `output_path` is more conventional for CLI tools.

lewis6991 replied:
Both are improvements over `file`. I still prefer `output_path` (or maybe `outpath`) using `path`, as `filepath` isn't a term I'm used to, or one we even use(?), it reads to me the same as `textstring`. Maybe go with `outfile`.

justinmk commented on runtime/lua/vim/net.lua:
```diff
@@ -0,0 +1,71 @@
+  else
+    vim.list_extend(args, { '--silent', '--show-error', '--fail' })
+  end
+  table.insert(args, '--location')
+  table.insert(args, '--retry')
+  table.insert(args, tostring(retry))
```
nit: list_extend ?

justinmk commented on runtime/lua/vim/net.lua:
```diff
@@ -0,0 +1,71 @@
+---
+--- @param url string The URL for the request.
+--- @param opts? table Optional parameters:
+---   - `verbose` (boolean|nil): Enables curl verbose output.
```
(not a blocker) Maybe this should be controlled by the 'verbose' option. Else we will need `verbose` flags on many functions. 

```suggestion
---   - `verbose` (boolean|nil): Enables verbose output.
```

justinmk replied:
```suggestion
---   - `verbose` (boolean|nil): Enables verbose output.
```

justinmk commented on runtime/doc/lua.txt:
```diff
@@ -4656,4 +4656,36 @@ tohtml.tohtml({winid}, {opt})                         *tohtml.tohtml.tohtml()*
+                       success.
+
+    Return (multiple): ~
+        (`string|boolean?`) In sync mode, returns the response body or `true`
```
see https://github.com/neovim/neovim/pull/34140/files#r2174793336 , should return a `{ text: string }` object rather than `string`.

lewis6991 commented on runtime/doc/lua.txt:
```diff
@@ -4656,4 +4656,36 @@ tohtml.tohtml({winid}, {opt})                         *tohtml.tohtml.tohtml()*
+                       • `output` (string|nil): A file path to save the
+                         response body to. If set, the success value will be
+                         `true` instead of the response body.
+      • {on_response}  (`fun(err?: string, content?: string|boolean)?`)
```
Honestly, it's probably wise to take inspiration from https://requests.readthedocs.io/en/latest/ for a lot of these API decisions.

justinmk replied:
yes, I've been referencing Go and nodejs fetch(), but sure that's good too.

justinmk commented on test/functional/lua/net_spec.lua:
```diff
@@ -0,0 +1,80 @@
+      local result
+      local M = require('vim.net')
+
+      M.request("https://httpbingo.org/anything", { retry = 3 }, function(err, body)
```
we will need to define new `NVIM_TEST_INTEG` env var and document it here: https://github.com/neovim/neovim/blob/f731766474901e5e345e0ca630315ef69122e556/test/README.md?plain=1#L436

It should be disabled by default. When enabled, it enables these tests which make network calls. Can use `t.skip()` to guard these tests, or define something like `skip_integ` similar to :  https://github.com/neovim/neovim/blob/f731766474901e5e345e0ca630315ef69122e556/test/testutil.lua#L837

tampueroc replied:
Noted!

justinmk commented on runtime/lua/vim/net.lua:
```diff
@@ -0,0 +1,82 @@
+
+  if opts.outpath then
+    table.insert(args, '--output')
+    table.insert(args, opts.outpath)
```
list_extend

justinmk commented on runtime/lua/vim/net.lua:
```diff
@@ -0,0 +1,81 @@
+    end
+  end
+
+  local job = vim.system(args, { text = true }, on_exit)
```
Is `text=true` always valid? What happens if a `.jpg` url is requested ?

Maybe we should drop `text=true` here, and return `{body: string}` and let the caller decide how to treat `body` for now. (Note: Lua strings can have binary data.)

justinmk commented on runtime/lua/vim/net.lua:
```diff
@@ -0,0 +1,81 @@
+
+  local function on_exit(res)
+    local err_msg = nil
+    local response_obj = nil
```
```suggestion
    local response = nil
```

tampueroc committed:
feat(net): add minimal vim.net.request() using curl

Problem:
Neovim has no simple HTTP GET API; netrw is too heavy for basic fetches.

Solution:
Add `vim.net.request()` as a thin curl wrapper:
  - Basic GET with --silent, --show-error, --fail, --location, --retry
  - Optional `opts.outpath` to save to a file
  - It operates asynchronously requiring an `on_response` handler to be
    passed when invoked.

Also included:
  - Functional tests for success and 404 failure (this tests are skipped based on
    NVIM_TEST_INTEG env var value)
  - Docs in `runtime/doc/lua.txt` under `vim.net.request()`
  - Health check for missing `curl`
  - Built in plugin that replaces `:edit` behaviour for editing remote
    files using the new `vim.net` module.
  - News entry in `doc/news.txt` under “LUA”

justinmk commented on runtime/doc/lua.txt:
```diff
@@ -5043,4 +5043,32 @@ tohtml.tohtml({winid}, {opt})                         *tohtml.tohtml.tohtml()*
+                       • `outpath` (string|nil): File path to save the
+                         response body to. If set, the `body` value in the
+                         Response Object will be `true` instead of the
+                         response body.
```
Something for "Next steps": when `outpath` is set, we could stream the response, so that e.g. a GB-sized download yields to the main thread on each ~chunk.

justinmk commented on runtime/lua/vim/net.lua:
```diff
@@ -0,0 +1,63 @@
+--- completetion. The `body` field in the response object contains the raw response data (text or binary).
+--- Called with (err, nil) on failure, or (nil, { body = string|boolean }) on success.
+function M.request(url, opts, on_response)
+  vim.validate({
```
this is the deprecated form of validate, now we prefer to do 3 validate() calls instead of 1 with a table.

justinmk commented on runtime/lua/vim/net.lua:
```diff
@@ -0,0 +1,63 @@
+
+    if res.code ~= 0 then
+      err_msg = (res.stderr ~= '' and res.stderr)
+        or string.format('Request failed with exit code %d', res.code)
```
followup: probably want a smart message, something that mentions that HTTP code if we got that far?

justinmk commented on runtime/plugin/nvim/net.lua:
```diff
@@ -0,0 +1,43 @@
+
+  local bufnr = args.buf
+  local url = vim.api.nvim_buf_get_name(bufnr)
+  local view = vim.fn.winsaveview()
```
Is this because it's async? We can try this. But maybe it makes sense to force it to be synchronous by using `vim.wait()`.

justinmk commented on runtime/plugin/nvim/net.lua:
```diff
@@ -0,0 +1,43 @@
+  local url = vim.api.nvim_buf_get_name(bufnr)
+  local view = vim.fn.winsaveview()
+
+  vim.api.nvim_echo({ { 'Fetching ' .. url .. ' …', 'MoreMsg' } }, true, {})
```
use `('...'):format()` almost always. using `..` to concatenate usually ends up being less readable.

justinmk commented on test/functional/lua/net_spec.lua:
```diff
@@ -0,0 +1,60 @@
+    n:clear()
+  end)
+
+  it('fetches a URL into memory (async success)', function()
```
let's also have a test that tests text response and binary response (jpg or something)

justinmk reviewed (approved):
The interface looks fine, so let's get this merged and we can keep iterating. Thanks for following through on this!

Tried `:edit https://...` locally, the "Fetching" message works nicely with "ui2".

Do you plan to close https://github.com/neovim/neovim/issues/7189 ?

tampueroc commented:
> The interface looks fine, so let's get this merged and we can keep iterating. Thanks for following through on this!
> 
> Tried `:edit https://...` locally, the "Fetching" message works nicely with "ui2".
> 
> Do you plan to close #7189 ?

Yes, happy to close it. Let me digest the feedback on this one first and I'm happy to tackle the follow-ups afterward.

michelesr commented:
> > @tampueroc After this change the filetype detection is not automatically applied anymore when using `:e <url>`, contrary to before when wget was used to download the file. In addition to this, the buffer is marked as unsaved (e.g. `[+]`) while it wasn't before, and reloading with `:e` gives error contrary to before when it would simply redownload a fresh version from the url.
> > P.S. Not saying it's a problem, as I'm not sure whether there are strict requirements on how the `:e` functionality should work with urls. Having the text higlighted and the right filetype detected was a nice to have. Another minor thing is that `:r` at the moment is not covered, so a similar autocmd could be added to handle that, for consistency.
> 
> Thank you for flagging this regressions! Are you happy to take them all? All of them are more or less one liners :)

I can give it a go on my spare time, I'm not sure I'll be able to do it so if in the meantime someone else wants to try is welcome. Are we aiming to replicate all those peculiarities that I listed or only a subset of them?
